pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyCRX.hpp

```diff
@@ -0,0 +1,289 @@
+        std::array<uint8_t, packed_size> arr{};
+
+        uint8_t s = (uint8_t{1U} << target);
+
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            arr[2 * k + 0] = 2 * (k ^ s) + 1;
+            arr[2 * k + 1] = 2 * (k ^ s) + 0;
+        }
+        return Permutation::compilePermutation<PrecisionT>(arr);
+    }
+
+    template <size_t target, typename ParamT>
+    static void applyExternalInternal(std::complex<PrecisionT> *arr,
+                                      size_t num_qubits, size_t control,
+                                      bool inverse, ParamT angle) {
+        // control qubit is external but target qubit is external
+        // const size_t rev_wire_min = std::min(rev_wire0, rev_wire1);
+        using namespace Permutation;
+
+        const size_t control_shift = (static_cast<size_t>(1U) << control);
+        const size_t max_wire_parity = fillTrailingOnes(control);
+        const size_t max_wire_parity_inv = fillLeadingOnes(control + 1);
+
+        constexpr static auto perm = applyExternalInternalOffDiagPerm<target>();
+
+        if (inverse) {
+            angle *= -1.0;
+        }
+        const auto diag_factor =
+            set1<PrecisionT, packed_size>(std::cos(angle / 2));
+        const auto offdiag_factor =
+            imagFactor<PrecisionT, packed_size>(-std::sin(angle / 2));
+
+        for (size_t k = 0; k < exp2(num_qubits - 1); k += packed_size / 2) {
+            const size_t i0 =
+                ((k << 1U) & max_wire_parity_inv) | (max_wire_parity & k);
+            const size_t i1 = i0 | control_shift;
+
+            const auto v1 =
+                PrecisionAVXConcept::load(arr + i1); // control bit is 1
+            const auto w1 = Permutation::permute<perm>(v1);
+            PrecisionAVXConcept::store(arr + i1,
+                                       diag_factor * v1 + offdiag_factor * w1);
+        }
+    }
+
+    template <typename ParamT>
+    static void applyExternalExternal(std::complex<PrecisionT> *arr,
+                                      const size_t num_qubits,
+                                      const size_t control, const size_t target,
+                                      bool inverse, ParamT angle) {
+        using namespace Permutation;
+        const size_t control_shift = static_cast<size_t>(1U) << control;
+        const size_t target_shift = static_cast<size_t>(1U) << target;
+
+        const size_t rev_wire_min = std::min(control, target);
+        const size_t rev_wire_max = std::max(control, target);
+
+        const size_t parity_low = fillTrailingOnes(rev_wire_min);
+        const size_t parity_high = fillLeadingOnes(rev_wire_max + 1);
+        const size_t parity_middle =
+            fillLeadingOnes(rev_wire_min + 1) & fillTrailingOnes(rev_wire_max);
+
+        if (inverse) {
+            angle *= -1.0;
+        }
+
+        const auto cos_factor =
+            set1<PrecisionT, packed_size>(std::cos(angle / 2));
```

Comment:
Could we use FP Broadcast operation here for the initialization?

Reply:
Currently, `set1` is defined to initialize AVX2/512 types directly. I will try broadcast functions. Still, I think I had a compiler error if I use AVX2/512 functions in a `constexpr` functions.

Reply:
Yeap, calling AVX intrinsics are not allowed in `constexpr` functions.